Sequencial issue index numbering with pessimistic locking mechanism

guillep2k opened the pull request:
As a stepping stone of my WIP #9787, I needed a pessimistic locking mechanism. Since this functionality is useful to solve many problems, I thought I'd use it for a simpler case to test the waters.

This PR introduces the pessimistic locking mechanism and finally gives proper solution to the problem from #7887 of calculating the index for two issues inserted at the very same time in the same repository.

The pessimistic locking is similar in concept to a mutex but works at the database level. This will allow synchronization between different Gitea instances that share the same database but run in different servers.

In addition, this PR uses the new locking mechanism to store the last issue index used for each repository in order to calculate the next value. This is a requirement for implementing the ability of deleting issues (#923), because we will want to avoid re-using the same issue number when one has been deleted (we're currently calculating the issue index by executing a `SELECT MAX(index)+1`, which will repeat values if issues are deleted from the end of the count).

### Background

Short explanation of the concurrency problem: https://github.com/go-gitea/gitea/issues/7887#issuecomment-537080059

Last year I tried to solve the potential problem of two threads attempting to create issues in the same repo at the same time: one of the transactions would fail because the other used the same issue index number. Reasons are too complex to detail here (see my attempts #7894, #7950, #7898, #8005, and the accepted makeshift solution in #8307), but ultimately the problem lays in the optimistic locking used in Gitea: two concurrent transactions will run unaware of the other until one of them commits, rendering the other transaction's calculations invalid.

### What this solves

Currently we simply retry the issue creation if the insert fails, and it should fail very rarely. However, the proper way to solve this is by effectively serializing the number assignment.

This PR provides the serialization mechanism so it can be used in more complex scenarios later.

zeripath commented on models/locked_resource.go:
```diff
@@ -0,0 +1,110 @@
+}
+
+// TempLockResource locks the given key but does not leave a permanent record
+func TempLockResource(e Engine, lockType string, lockKey int64) error {
```
This is basically a (key)notexists or (key)isavailable test. That might be a better name depending on how you were intending on using this.

guillep2k replied:
The intention is however to _lock_ a type of resource. The "(key)notexists functionality" is only a convenience to help preventing a bad usage. Maybe I should  rename it to something_something_Mutex?

guillep2k replied:
Changed the name to a more informative but verbose `TemporarilyLockResourceKey`.

zeripath commented:
Hmm thinking on the lock process should provide a DBContext itself.

guillep2k commented:
> 
> 
> Hmm thinking on the lock process should probably provide a DBContext itself.

You mean something like (function names aside)...?:

```
WithResourceLocked("my-type",444, func(){
    ... do something ...
})

zeripath commented:
Something like that - I've not fleshed it out in my brain though. 

Callbacks aren't greatly favoured in the Go world - for example, the lack of a `io.With( ... )` function or `with` keyword - with preference for defers but I think there's definitely a place for them.

It might be that a specific LockedDBContext or something like that is better.

guillep2k commented:
> 
> 
> Something like that - I've not fleshed it out in my brain though.
> 
> It might be that a specific LockedDBContext or something like that is better.

It's not that these functions need a `defer`, because the update will be effective on commit (or cleaned up on rollback). Anyway, this kind of functions can easily be added later without breaking anything, so we can wait for more use cases to decide.

guillep2k commented:
@zeripath done. I've reworked the interface a little and added support for `LockKey == 0` in case it doesn't make sense to pick a particular value for it.

lunny commented on models/locked_resource.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+// UpdateValue updates the value of the counter of a locked resource
+func (r *LockedResource) UpdateValue() error {
```
How did we lock the counter?

guillep2k replied:
> How did we lock the counter?

@lunny at the upsert in `GetLockedResource`. Being an insert/update, the record remains locked for the rest of the transaction. The counter is a column of the `locked_resource` table.

lunny replied:
The transaction in fact will not lock your record I think.

guillep2k replied:
> 
> 
> The transaction in fact will not lock your record I think.

@lunny I've written tests to prove that 😁. Check [`models/integrations/locked_resource_test.go`](https://github.com/guillep2k/gitea/blob/bylock-indexes/integrations/locked_resource_test.go).

lunny commented:
And why not to use an optimistic lock but a pessimistic lock? The record seems not be locked when select ?

guillep2k commented:
> 
> 
> And why not to use an optimistic lock but a pessimistic lock? The record seems not be locked when select ?

@lunny See my explanation of the concurrency problem with optimistic locks: #7887 (comment)

The records are not locked on SELECT unless it's a SELECT FOR UPDATE, but that doesn't work for _new_ records (INSERT) in all of our supported databases.

This PR solves the index problem in issue, but more than anything it provides a standarized way of using lock on operations that otherwise could get wrong results (as it's happening now with the issue index calculation, that why we need to retry if it fails). As you can see, with a pessimistic lock I could get rid of the retries in that case.

I've tried many combinations (SELECT ... FOR UPDATE, INSERT, SELECT ... FOR UPDATE then INSERT, etc.), but UPSERT was the only method that passed all the tests.

lunny commented:
@guillep2k As I know, transaction will not lock any record. So I don't know how this will lock the record.

lafriks commented:
@lunny if you update record it will be locked for other updates until previous transaction is commited

lunny commented:
@lafriks Yes, but if both two creating issue requests. The sequence maybe below, but I think the result is still wrong?
```
first get
second get
second update
first update
```

6543 commented:
I'm just thinking if using the cache give us **more speed** and you also can do this multi-inctance if you use the right cache backend

guillep2k commented:
@lunny I've manually tested this on each DB and also created the integration tests to automate this check, so I'm pretty confident that this works. In the past we had problems because we did `SELECT`, then `INSERT`; that won't work because there is a gap between the first and second statements and data can change in the meantime. In this case I use `UPSERT` that definitively locks the key until the end of the transaction. The other important detail here is that I use a single row for the task instead of a `SELECT MAX(*) ... FOR UPDATE`, which doesn't exist, and if it would, it would lock the entire`item` table, which is unacceptable.

6543 commented on models/locked_resource.go:
```diff
@@ -0,0 +1,121 @@
+// the function will not return until it acquires the lock or receives an error.
+func upsertLockedResource(e Engine, resource *LockedResource) (err error) {
+	// An atomic UPSERT operation (INSERT/UPDATE) is the only operation
+	// that ensures that the key is actually locked.
```
this should go in xorm .. a atomic InsertUpdate func ... This is needed in many places

guillep2k replied:
The reason I didn't go down that path is that I believe we need pessimistic locking in quite a few places, not just this insert.

6543 commented:
i know time is less ...

lunny commented on models/locked_resource.go:
```diff
@@ -0,0 +1,121 @@
+	LockKey  int64  `xorm:"pk"`
+	Counter  int64  `xorm:"NOT NULL DEFAULT 0"`
+
+	engine Engine `xorm:"-"`
```
Could we still use engine as a parameter so that we can find the error obviously as before?

guillep2k replied:
I think this would have prevented errors instead (e.g. people calling on `x` would cause problems). I changed it, though.

lunny commented on models/locked_resource_test.go:
```diff
@@ -0,0 +1,101 @@
+func TestLockedResource(t *testing.T) {
+	assert.NoError(t, PrepareTestDatabase())
+
+	withSession := func(t *testing.T, f func(t *testing.T, sess *xorm.Session) bool) {
```
rename it as `withTransaction` maybe better.

guillep2k commented:
@lunny done.

lunny commented:
Closed per #15599
